src/category_theory/sites/closed.lean

```diff
@@ -0,0 +1,298 @@
+and further this presheaf can be used to determine the Grothendieck topology from the sheaf
+predicate.
+Finally we show that a natural closure operator on sieves induces a Grothendieck topology, and hence
+that natural closure operators are in bijection with Grothendieck topologies.
+
+## Tags
+
+closed sieve, closure, Grothendieck topology
+
+## References
+
+* [S. MacLane, I. Moerdijk, *Sheaves in Geometry and Logic*][MM92]
+-/
+
+universes v u
+
+namespace category_theory
+
+variables {C : Type u} [small_category C]
+
+variables {P : Cᵒᵖ ⥤ Type u}
+variables (J₁ J₂ : grothendieck_topology C)
+
+/-- The `J`-closure of a sieve is the collection of arrows which it covers. -/
+@[simps]
+def grothendieck_topology.close {X : C} (S : sieve X) : sieve X :=
+{ arrows := λ Y f, J₁.covers S f,
+  downward_closed' := λ Y Z f hS, J₁.arrow_stable _ _ hS }
+
+/-- Any sieve is smaller than its closure. -/
+lemma grothendieck_topology.le_close {X : C} (S : sieve X) : S ≤ J₁.close S :=
+λ Y g hg, J₁.covering_of_eq_top (S.pullback_eq_top_of_mem hg)
+
+/--
+A sieve is closed for the Grothendieck topology if it contains every arrow it covers.
+In the case of the usual topology on a topological space, this means that the open cover contains
+every open set which it covers.
+
+Note this has no relation to a closed subset of a topological space.
+-/
+def grothendieck_topology.is_closed {X : C} (S : sieve X) : Prop :=
+∀ ⦃Y : C⦄ (f : Y ⟶ X), J₁.covers S f → S f
+
+/-- If `S` is `J₁`-closed, then `S` covers exactly the arrows it contains. -/
+lemma grothendieck_topology.covers_iff_mem_of_closed {X : C} {S : sieve X}
+  (h : J₁.is_closed S) {Y : C} (f : Y ⟶ X) :
+  J₁.covers S f ↔ S f :=
+⟨h _, J₁.arrow_max _ _⟩
+
+/-- Being `J`-closed is stable under pullback. -/
+lemma grothendieck_topology.is_closed_pullback {X Y : C} (f : Y ⟶ X) (S : sieve X) :
+  J₁.is_closed S → J₁.is_closed (S.pullback f) :=
+λ hS Z g hg, hS (g ≫ f) (by rwa [J₁.covers_iff, sieve.pullback_comp])
+
+/--
+The closure of a sieve `S` is the largest closed sieve which contains `S` (justifying the name
+"closure").
+-/
+lemma grothendieck_topology.le_close_of_is_closed {X : C} {S T : sieve X}
+  (h : S ≤ T) (hT : J₁.is_closed T) :
+  J₁.close S ≤ T :=
+λ Y f hf, hT _ (J₁.superset_covering (sieve.pullback_monotone f h) hf)
+
+/-- The closure of a sieve is closed. -/
+lemma grothendieck_topology.close_is_closed {X : C} (S : sieve X) : J₁.is_closed (J₁.close S) :=
+λ Y g hg, J₁.arrow_trans g _ S hg (λ Z h hS, hS)
+
+/-- The sieve `S` is closed iff its closure is equal to itself. -/
+lemma grothendieck_topology.is_closed_iff_close_eq_self {X : C} (S : sieve X) :
+  J₁.is_closed S ↔ J₁.close S = S :=
+begin
+  split,
+  { intro h,
+    apply le_antisymm,
+    { intros Y f hf,
+      rw ← J₁.covers_iff_mem_of_closed h,
+      apply hf },
+    { apply J₁.le_close } },
+  { intro e,
+    rw ← e,
+    apply J₁.close_is_closed }
+end
+
```

Comment:
Do you want a simp lemma `close_close`? And maybe a variant of the `iff` above that says `J.close S = S` assume `J.is_closed`? I assume those might be useful for rewriting later on.